README.md
# ***pyTableMaker***
**The library / module for python to allow programmers to create, edit and show tables without gui conveniently.**  
[![Run on Repl.it](https://repl.it/badge/github/windowsboy111/pyTableMaker)](https://repl.it/github/windowsboy111/pyTableMaker)

> Check out changelog at CHANGELOG.md  
> run example.py to check out how it performs!

---

# Info
There are 2 *editions* of pyTableMaker:
1. the one on github
2. the one on pypi

These 2 editions have difference between when importing, but other than that, there's not much difference.

Syntax of importing for **github edition**:
```py
import table
myTable = table.modernTable()
```
Syntax of importing for **pypi edition**:
```py
import pyTablemaker
myTable = pyTableMaker.modernTable()
```
Pretty simple, right?

---

# Usage
## Types of table
### modernTable
- double line, probably the most perfect-looking table
- formed with ╔ ═ ╗ ╦ ╠ ╣ ╬ ╚ ╝ ╩ ║
### classicTable
- for consoles with ASCII support only
- formed with + | -
### onelineTable
- similar to *modernTable* but with one line only, self explainatory
- formed with ┌ ─ ┐ ┬ ├ ┤ ┼ └ ┘ ┴ │
### customTable
- this is the base class of all other table classes (aka other are abstract)
- you can customize your table how it is formed using this class, self explainatory again
## creating a table
```py
t = modernTable(align='right',exp=True,linespacing=1) # please don't do linespacing it's ugly unless you need to
```
### parameters
- align="left" : str {"left","center","right"} (align the text for each column)
- exp=False : bool                             (Express mode allows you to stack up methods, see below ￬ )
- linespacing=0 : int                          (add a row on top and below a item / value)
### return
- table object
## express mode
You can turn on express mode when you initialize / declare the table. ￪ see above  
After turning on, you can stack up methods, and calling them once. all methods will return `self` in express mode except `get()`.
```py
t2 = classicTable(exp=True).new_column('id').table.new_column('name').table.new_column('price').table.new_column('note').rename('remarks').table\
    .new_column('description').table\
    .insert(0,'coke',          'A popular drink in brown color',   3.5,   'Need more, ppl keep buying, keep running out!')\
    .insert(1,'chips',         None,                               0xffff,"It comes from the space lol idk.")\
    .insert(2,'tea',           'Hot green tea i guess.',           5,     None)\
    .insert(3,'Cup noodles',   None,                               40,    'Need many cuz Corona Virus.')\
    .show(2) # you can specify how many rows to get as a argument. (0 is ignored) see below ￬
#   ~^~~~~~~  just a side note that there's a show() method so you don't need to do print(t.get()) ￬ see below.
```
## append a column
```py
col = t.new_column(name)
```
### parameters
- name : str
### return
- column object linked to the table object
## delete a column
```py
col.delete()
```
## rename a column
```py
col.rename(newColumnName)
```
### parameters
- newColumnName : str
## move a column
You can only move a column to the end (the most right hand side) of a table.
```py
col.moveToEnd()
```
## insert a row
```py
t.insert(v1,v2,v3)
t.insert(values)
```
### parameters
- v1, v2, v3... : any
- values : tuple
## print the table
```py
t.show(rowCount)
```
### parameters
- rowCount=0 : int (how many rows you want, default is 0 which shows all)
## get the table in string form
**IMPORTANT: t.get() method will always return the string form of the table even express mode is turned on!**
```py
result = t.get(rowCount)
```
### parameters
- rowCount=0 : int (how many rows you want, default is 0 which returns all)
### return
- result : str  
note: you can print the string form of table using `t.show()`. see above ￪
## number of rows in a table
```py
t.row
```

CHANGELOG.md
# Changelog of pyTableMaker
    windowsboy111

---
## [Unreleased]
### Fixed
- when forming table, all columns will use the first column column length to count how many blank spaces to leave (forget to `colNum += 1` so it always uses `colMaxLen[0]`)

## [OOP Update Unreleased]
### Added
- support for `+`, `-`, `len()`, comparing operators such as `<`, `<=`, `!=`
- specifying return type
### Fixed
- PEP formatting

# [1.3-a2]
### Added
- cell wraping (change it in settings / kwargs)
### Removed
- revert all syntax / classname changes for last master prl

## [1.3-a1] - 2020-07-10 [YANKED]
### Changed
- Table classes now looks like this: `CustomTable`, `ModernTable` instead of `camelCase`
### Fixed
- `.copy()` now clone a table to `CustomTable`
- following syntax rules (PEP)

## [1.3-a] - 2020-06-22
### Added
- you can now show limited rows with `get(n)` or `show(n)` in which `n` is the number of rows you want.
### Fixed
- `insert()` did not return `colMaxLen` in normal mode.
- `remove()` fixed.
### Changed
- `dummy_func()` is no longer used. Instead we have `# pylint: disable=unused-variable`.
### Removed
- `OrderedDict()`. Bruh time! Last update I forgot to remove it for some reason...

## [1.2.1] - 2020-05-16
### Added
- readme added run on repl.it badge.
- changelog check changes section.
- Avaliable on pypi!
### Removed
- `dict()` has been replaced.

## [1.2.0] - 2020-05-03
### Added
- changelog which is what you are looking at rn.
- new type of table: `customTable` that allows user / programmer to define what characters the table can be made of.
- now kwargs can be passed to the constructor of the table class for different settings. If an invalid setting is passed, `invalidSettings` expection will be raised.
- new setting: `align` that specify how text in cells should align such as left, right and center.
- new setting: `linespacing` that specify how many lines cells should add. default is 0. (not recommened as it is ugly unless you need some breathing space)
- new setting: `exp` aka express mode. When set to `True`, methods can stack up each other.
- new method in class `column`: `getTable()` which returns the pointer of the parent class.
- new method: `show()` which is an alias of `print(t.get())`
- new global variable in library: `lib-info` that stores info about the library.
- added file: table_test.py to test the library.
### Changed
- the instance variable `obj` in class `column` has been renamed to `table` to make it more clear that it stores the `self` pointer of its parent class.
- updated `example.py` as expected to show new features.
- `customTable` is now the parent of all table classes.
- Running `table.py` directly (i.e. `python3 table.py`) will print `lib-info` instead of doing nothing.
- Reorganized the code, making `get()` less clutter / simplified the code so that it no longer violates the ultimate DRY rule:
> - `__init__()` ==> `_init()` and `__init__()`              (Not recommend to call them directly)
> - `get()`      ==> `_get()`, `_get_l()` and `get()`        (Not recommend to call `_get()` and `_get_l()` directly)
### Fixed
- no more warnings when using the library, `dummy_func()` has been added to prevent unused variable warnings.
### Deprecated
- soon in a future update, `OrderedDict()` will be replaced with normal standard `dict()`. (Backwards compatible warning!)

## [1.1.0] - 2020-04-13
### Added
- new type of table: `onelineTable`, similar to `modernTable`, but have only one border instead of double borders.

## [1.0.0] - 2020-04-10
### Added
- new type of table: `classicTable` that is formed with +, -, and |.
- Columns can now be renamed with `column.rename(newName)`.
- Columns can now be moved to the end with `column.moveToEnd()`.
- Columns can now be deleted with `column.delete(newName)`.
### Changed
- the original table type now has been renamed to `modernTable`.
- Columns needs to be added with `table.new_column(columnName)` instead of method `add_column()`.
- `new_column()` now returns an column object instead of nothing.
### Fixed
- `OrderedDict` has been used to store data instead of `dict`. This is being considered as a bug because before python 3.7, the standard `dict` type does NOT maintain insertion order by default.

## [0.1.0] - 2020-04-09
### Added
- example for the library.
- a basic table class.

[OOP Update Unreleased]:    https://github.com/windowsboy111/pyTableMaker/compare/OOP-update
[Unreleased]:   https://github.com/windowsboy111/pyTableMaker/compare/1.3-a2...HEAD
[1.3-a2]:       https://github.com/windowsboy111/pyTableMaker/compare/1.3-a1...1.3-a2
[1.3-a1]:       https://github.com/windowsboy111/pyTableMaker/compare/1.3-a...1.3-a1
[1.3-a]:        https://github.com/windowsboy111/pyTableMaker/compare/1.2.1...1.3-a
[1.2.1]:        https://github.com/windowsboy111/pyTableMaker/compare/v1.2.0...1.2.1
[1.2.0]:        https://github.com/windowsboy111/pyTableMaker/compare/v1.1.0...v1.2.0
[1.1.0]:        https://github.com/windowsboy111/pyTableMaker/compare/v1.0.0...v1.1.0
[1.0.0]:        https://github.com/windowsboy111/pyTableMaker/compare/v0.1.0...v1.0.0
[0.1.0]:        https://github.com/windowsboy111/pyTableMaker/releases/tag/v0.1.0


test.py
from table import modernTable


t = modernTable()
col = t.new_column('hi bro what\'s up? ha ha ha!')
col1 = t.new_column('ahhhhhhh :)')
t.insert('ahhhhhhhhhhhhh', "yes you get it right, finally!")
t.insert('a', 34567)
t.show()


table_test.py
from table import customTable
print(customTable(['+','-','+','+','+','-','+','+','+','-','+','+','|'],align='center',linespacing=1,exp=True) \
    .new_column('column 1').table.new_column('column 2').table.insert('row 1','col 2').get())

table.py
DEFAULT_SETTINGS = {
    "align":        "left",
    "linespacing":  0,
    "exp":          False,
    "cellwrap":     40,
    "linewrap":     150  # not implemented >w<
}


class invalidSettings(Exception):
    """The settings passed via kwargs from the constructor of the class is invalid"""

    def __init__(self, message, possible=None, reason=""):
        super.__init__(super(str, message))
        self.message = message
        self.possible = possible
        self.reason = reason

    def __str__(self):
        return self.message

    def __repr__(self):
        return {'error': invalidSettings, 'message': self.message, 'possible': self.possible, 'reason': self.reason}


class customTable:
    def __repr__(self):
        return str(self.data)

    def __str___(self):
        return 'customTable'

    def __init__(self, style: list, data=None, colMaxLen: list = [], **kwargs):
        """\
        Initialize a table creation.\n
        Required arg1: style. type is list. charaters to form table.\n
        Optional arg1: data.  type is dict.\n
        Optional arg2: colMaxLen.  type is list.  Must be passed if you have the dict (arg1) passed.\
        """
        self._init(style, data, colMaxLen, **kwargs)

    def _init(self, style: list, data=None, colMaxLen: list = [], **kwargs):
        self.style = style
        self.data = data or dict()
        self.row = 0
        self.colMaxLen = colMaxLen
        self.settings = DEFAULT_SETTINGS.copy()
        self.settings.update(kwargs)

    def copy(self):
        return customTable(self.style, self.data, self.colMaxLen, **self.settings)

    def new_column(self, name):
        class column:
            def __init__(self, table, name):
                """\
                create a new column (initialize)\n
                arg1: name.  Name of the new column.\n
                return: column location\
                """
                table.data[name] = list()
                table.colMaxLen.append(len(name))
                self.name = name
                self.location = len(table.colMaxLen) - 1
                self.table = table

            def rename(self, newname):
                """
                rename the column (and move to end)
                arg1: newname.
                return: old name
                """
                oldname = self.name
                rowdata = self.table.data[self.name]
                del self.table.data[self.name]
                self.table.data[newname] = rowdata
                self.name = newname
                self.location = len(self.table.colMaxLen) - 1
                if self.table.settings['exp']:
                    return self
                return oldname

            def delete(self):
                """
                delete the column
                no arguments
                """
                del self.table.data[self.name]
                if self.table.settings['exp']:
                    return self

            def move_to_end(self):
                """
                move a column
                no arguments
                """
                self.table.data[self.name] = self.table.data.pop(self.name)
                self.location = len(self.table.colMaxLen) - 1
                if self.table.settings['exp']:
                    return self

            def get_table(self):
                return self.table
        return column(self, name)

    def insert(self, *values):
        """
        insert a row.
        arg1: values. The value for the row.
        return: the longest length of columns (list)
        """
        self.row += 1
        # for the index loop of the list
        loopCount = 0
        # column is the key
        for column in self.data:
            self.data[column].append(str(values[loopCount]))
            self.colMaxLen[loopCount] = max(self.colMaxLen[loopCount],
                                            len(str(values[loopCount])))
            loopCount += 1
        if self.settings['exp']:
            return self
        return self.colMaxLen

    def remove(self, index):
        """
        remove a row.
        arg1: index. count from 0. rowNum.
        return: removed items / values
        """
        result = list()
        self.row -= 1
        loopCount = 0
        for column in self.data:
            result.append(self.data[column].pop(index))
            self.colMaxLen[loopCount] = max(self.data[column])
            loopCount += 1
        if self.settings['exp']:
            return self
        return result

    def get(self, rq=0):
        """Return back the table in string.  No arguments."""
        return self._get(rq)

    def _add_linespacing(self, s12):
        result = ''
        for i in range(self.settings['linespacing']):
            for colNum in range(len(self.data)):
                result += f'{s12} ' + ' ' * \
                    (self.colMaxLen[colNum] if self.colMaxLen[colNum] <= (self.settings['cellwrap'] - 2) else (self.settings['cellwrap'] - 2)) + ' '
            result += s12 + '\n'
        return result

    def _get_l(self, s12, rowNum=-1):
        result = self._add_linespacing(s12)

        colNum = 0  # get loop count ready
        nextVal = list()  # storing next line to be placed if cell / line wrapping is needed
        maxLen = self.settings['cellwrap'] - 2
        for k, v in self.data.items():
            name = k if rowNum == -1 else v[rowNum]
            if len(name) > maxLen:
                nextVal.append(name[maxLen:])
                name = name[:maxLen]
            else:
                nextVal.append('')
            neededSpaces = (
                self.colMaxLen[colNum] if self.colMaxLen[colNum] <= maxLen else maxLen) - len(name)
            result += f'{s12} {name} ' + ' ' * neededSpaces
            colNum += 1
        result = result[:-1] + ' ' + s12 + '\n'
        while any(nextVal):
            loop = 0
            theNext = list()
            for name in nextVal:
                if len(name) > maxLen:
                    theNext.append(name[maxLen:])
                    name = name[:maxLen]
                else:
                    theNext.append('')
                neededSpaces = (
                    self.colMaxLen[colNum] if self.colMaxLen[colNum] <= maxLen else maxLen) - len(name)
                result += f'{s12} {name} ' + ' ' * neededSpaces
                loop += 1
            result = result[:-1] + ' ' + s12 + '\n'
            if theNext == ['' for col in self.data]:
                break
            nextVal = theNext.copy()

        result += self._add_linespacing(s12)
        return result

    def _get_struct(self, styleLeft, styleMid, styleCross, styleRight, colNum):
        result = styleLeft
        for column in self.data:
            result += styleMid * ((self.colMaxLen[colNum] + 2) if (
                self.colMaxLen[colNum] + 2) < self.settings['cellwrap'] else self.settings['cellwrap'])
            result += styleCross
            colNum += 1
        return result[:-1] + styleRight + '\n'

    def _get(self, rq):
        rq = rq or self.row
        colNum = 0
        result = self._get_struct(self.style[0], self.style[1], self.style[3], self.style[2], colNum)  # get line 1
        # 2nd line loop
        result += self._get_l(self.style[12])
        # every row and every top bar
        i = 0
        for rowNum in range(self.row):
            colNum = 0
            # the top bar
            result += self._get_struct(self.style[4], self.style[5], self.style[7], self.style[6], colNum)
            # the value
            result += self._get_l(self.style[12], rowNum)
            i += 1
            if i >= rq:
                break
        # finish the bottom of the table
        colNum = 0
        result += self._get_struct(self.style[8], self.style[9], self.style[11], self.style[10], colNum)
        return result

    def show(self, rq=0):
        print(self.get(rq))
        if self.settings['exp']:
            return self


class modernTable(customTable):
    def __str___(self):
        return 'modernTable'

    def __init__(self, data=None, colMaxLen: list = [], **kwargs):
        """\
        Initialize a Modern Table creation.\
        """
        self._init(['╔', '═', '╗', '╦', '╠', '═', '╣', '╬', '╚',
                    '═', '╝', '╩', '║'], data, colMaxLen, **kwargs)

    def get(self, rq=0):
        return self._get(rq)


class classicTable(customTable):
    def __str__(self):
        return 'classicTable'

    def __init__(self, data=None, colMaxLen: list = [], **kwargs):
        """\
        Initialize a classic table creation, which is formed with + | and -.\
        """
        self._init(['+', '-', '+', '+', '+', '-', '+', '+', '+',
                    '-', '+', '+', '|'], data, colMaxLen, **kwargs)

    def get(self, rq=0):
        """Return back the table in string.  No arguments."""
        return self._get(rq)


class onelineTable(customTable):
    def __str__(self):
        return 'onelineTable'

    def __init__(self, data=None, colMaxLen: list = [], **kwargs):
        """\
        Initialize a modern table creation, but have one border instead of double borders.\
        """
        self._init(['┌', '─', '┐', '┬', '├', '─', '┤', '┼', '└', '─', '┘', '┴', '│'], data, colMaxLen, **kwargs)

    def get(self, rq=0):
        return self._get(rq)


lib_info = """\
pyTableMaker by windowsboy111
Check out the repo:    https://github.copm/windowsboy111/pyTableMaker/
Here's the website:    https://windowsboy111.github.io/pyTableMaker/
Thanks for choosing this library! Remember to share it too!\
"""

if __name__ == "__main__":
    print(lib_info)


example.py
from table import *  # pylint: disable=unused-wildcard-import
t1 = modernTable()                       # init new table t (in this situation the type is modernTable)
col1 = t1.new_column('col1')             # add column named "col1" to t
col2 = t1.new_column('col2')             # this method will return a column object
col3 = t1.new_column('col3')             # to get the table of the column, use col.table or col.getTable()
col1.delete()                            # self-explainatory
col2.rename('wow')                       # rename column and move it to the end
col2.moveToEnd()                         # self-explainatory
t1.insert('a', 1.1235)                    # insert new row with values
t1.insert('bc', 3.14159)                  # Specifying the type of values is not needed
t1.insert('d', 9.00001)                   # all values will be converted to string
print(t1.get())                          # get table t in String and print out
print(t1.data)                           # get the raw dictionary of the table.
print(t1.row)                            # number of rows


# express mode (exp for short) is a mode that allows you to stack all the methods into a single statement.
# to enable express mode, add a kwargs called 'exp' and set it to True:
t2 = classicTable(align='right', exp=True)\
    .new_column('id').table\
    .new_column('name').getTable()\
    .new_column('price').table\
    .new_column('note').rename('remarks').table\
    .new_column('description').table\
    .insert(0, 'coke',          'A popular drink in brown color',   3.5,    'Need more, ppl keep buying, keep running out!')\
    .insert(1, 'chips',         None,                               0xffff, "It comes from the space lol idk.")\
    .insert(2, 'tea',           'Hot green tea i guess.',           5,      None)\
    .insert(3, 'Cup noodles',   None,                               40,     'Need many cuz Corona Virus.')\
    .show(2)  # you can specify how many rows to get as a argument. (0 is ignored)
#   ~^~~~~~~  just a side note that there's a show() method so you don't need to do print(t.get())
print(t2)
print(repr(t2))


t3 = onelineTable(align='center')                   # by default align is 'left', you can specify it with 'right' or 'center'
time = t3.new_column('Time')
event = t3.new_column('Event')
remarks = t3.new_column('Remarks')

remarks.rename('Note')
t3.insert('11:00', 'Brunch', 'I ate a cookie and a cup noodle which costs 40 dollars to buy at home.')
t3.insert('13:00', 'Coding', 'That\'s my hobby, right?')
t3.insert('15:55', 'Do hw',  'Annoying mum asks me to do hw :(')
t3.insert('19:30', 'Dinner', 'Nice curry ;)')
print(t3.get())
print()
print(lib_info)                                     # info about the library ;)
